### ISSUE TITLE
Unreliable filtering of interfaces usage in UseInterfaceOverImplementationInConstructorRector

### ISSUE THREAD

**Issue by SilverFire** [rectorphp/rector#4049]
The following code is NOT reliable and does NOT filter out interfaces, until it's multiple implementations ARE autoloaded.

https://github.com/rectorphp/rector/blob/e7429d96a4e277665a6ff55f23ea96902c74f3ab/rules/solid/src/Rector/ClassMethod/UseInterfaceOverImplementationInConstructorRector.php#L134-L160

There is [a test](https://github.com/rectorphp/rector/blob/master/rules/solid/tests/Rector/ClassMethod/UseInterfaceOverImplementationInConstructorRector/Fixture/skip_class_that_implements_interface_with_multiple_children.php.inc#L7-L7) for it, but it passes only because [the second implementation](https://github.com/rectorphp/rector/blob/5f4a2c1e3b735aa3ad5fa02ad3b34ca29793d924/rules/solid/tests/Rector/ClassMethod/UseInterfaceOverImplementationInConstructorRector/Source/Coconut.php#L12) of the [Fruit interface](https://github.com/rectorphp/rector/blob/master/rules/solid/tests/Rector/ClassMethod/UseInterfaceOverImplementationInConstructorRector/Source/Fruit.php#L7) is [manually added in the composer file-autoload](https://github.com/rectorphp/rector/blob/4553ae9579929b1cf9fa6a18c9ef932022375598/composer.json#L269). Witch, obviously, makes no sense IRL.

I don't think we should autoload all the files from the whole project in order to make this rule work, should we or there is other way to do it?

**Comment by TomasVotruba** [comment]
> I don't think we should autoload all the files from the whole project in order to make this rule work, should we or there is other way to do it?

Many rules requires this, like dead public constant or dead public methods. Without full project autoloaded, these rules can't work by definition.

If this rules causes you more troubles then value in your specific code base, it's should be excluded.

In long term value of this rule, how could we improve it to bring more value and help you with your code?

**Comment by SilverFire** [comment]
Do you know any shortcuts on how to autoload everything in a project?

**Comment by TomasVotruba** [comment]
composer > psr-4 > classmap is one of hacks, but can break something else

**Comment by SilverFire** [comment]
Nah, that's a bad idea

**Comment by TomasVotruba** [comment]
I thought so. 

Could you maybe describe the bug? I'm not sure I understand the problem here

**Comment by SilverFire** [comment]
I tried running this Rector in my codebase, where the particular implementation was used on purpose in the constructor. Rector has replaced the particular implementation dependency with the interface, regardless of the codebase contains multiple implementations of the interface.

I'm not sure what is the order of execution in Rector. I guess these kinds of rules may work correctly only when Rector:

1. Crawls the whole codebase and tries to load all possible classes/functions/etc 
2. Starts the analysis, having everything loaded and parsed

Does it work like this? From what I saw, I guess that it's loading the class being, analyzed, and everything that is used in it (recursively). But not the whole codebase.

Am I right?
